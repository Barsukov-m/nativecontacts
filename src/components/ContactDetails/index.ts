import Header from './Header';
import Section from './Section';
import ContactInfo from './ContactInfo';
import styles from './styles';

export { Header, Section, ContactInfo, styles };
