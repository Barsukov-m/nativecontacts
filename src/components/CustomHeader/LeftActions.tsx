import {
	NavigationProp,
	ParamListBase,
	useNavigation,
	useRoute,
} from '@react-navigation/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import styles from './styles';
import themeColors from '../../constants/themeColors';

export interface NavigationParams extends NavigationProp<ParamListBase> {
	goBack: () => void;
	openDrawer: () => void;
}

const LeftActions: React.FC = () => {
	const { name } = useRoute();
	const { goBack, openDrawer } = useNavigation<NavigationParams>();

	if (name !== 'Home') {
		return (
			<Icon
				onPress={() => goBack()}
				name="arrow-back"
				size={20}
				style={{ ...styles.icon, ...styles.leftAction }}
			/>
		);
	}

	return (
		<Icon
			onPress={() => openDrawer()}
			name="menu"
			size={20}
			style={{
				...styles.icon,
				...styles.leftAction,
				color: themeColors.primaryText,
			}}
		/>
	);
};

export default LeftActions;
