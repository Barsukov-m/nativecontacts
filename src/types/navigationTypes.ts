import { NativeStackNavigationProp } from '@react-navigation/native-stack';

export interface ScreenProps {
	navigation: NativeStackNavigationProp<any>;
}

export interface ContactInterface {
	id?: number;
	firstName: string;
	lastName: string;
	maidenName?: string;
	age?: number;
	gender?: string;
	email?: string;
	phone: string;
	username?: string;
	birthDate?: string;
	image?: string;
	address?: {
		address?: string;
		city: string;
		state?: string;
	};
	university?: string;
	company?: {
		name: string;
	};
}
