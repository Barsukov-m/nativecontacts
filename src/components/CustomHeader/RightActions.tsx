import { View, TouchableNativeFeedback } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { useRoute } from '@react-navigation/native';
import { ContactsRouteParams } from '../../types/navigationTypes';
import styles from './styles';

interface RightActionProps {
	onEdit: () => void;
	onDelete: () => void;
}

const RightActions: React.FC<RightActionProps> = ({ onEdit, onDelete }) => {
	const { params } = useRoute<ContactsRouteParams>();

	const contact = params?.contact;

	if (!contact) {
		return null;
	}

	return (
		<View style={styles.actionWrapper}>
			<TouchableNativeFeedback onPress={onEdit}>
				<View>
					<Icon
						name="mode"
						size={20}
						style={{
							...styles.icon,
							...styles.editBtn,
						}}
						testID="edit-button"
					/>
				</View>
			</TouchableNativeFeedback>
			{contact.id !== 0 && (
				<TouchableNativeFeedback onPress={onDelete}>
					<View>
						<Icon
							name="delete"
							size={25}
							style={{
								...styles.icon,
								...styles.editBtn,
							}}
							testID="delete-button"
						/>
					</View>
				</TouchableNativeFeedback>
			)}
		</View>
	);
};

export default RightActions;
