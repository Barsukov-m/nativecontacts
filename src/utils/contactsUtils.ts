import { Alert } from 'react-native';
import { ContactInterface } from '../types/navigationTypes';

export const capitalizeFirstLetter = (text: string): string => {
	return text[0].toUpperCase() + text.slice(1);
};

export const getHeaderTitleByContact = (contact: ContactInterface) => {
	const firstName = contact.firstName
		? capitalizeFirstLetter(contact.firstName)
		: '';
	const lastName = contact.lastName
		? capitalizeFirstLetter(contact.lastName)
		: '';
	return `${firstName} ${lastName}`;
};

export const logWarning = (title: string, message: string) => {
	Alert.alert(title, message);
};

export const wipWarning = () => {
	logWarning('Work in Progress ğŸ¤«', 'This feature is currently in development');
};
