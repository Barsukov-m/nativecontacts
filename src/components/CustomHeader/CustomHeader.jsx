import { TouchableNativeFeedback } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { wipWarning } from '../../utils/contacts';
import colors from '../../constants/themeColors';
import styles from './styles';

const CustomHeader = ({ navigation, route, title }) => {
  const handleContactEdit = () => {
    wipWarning();
  };

  const getLeftAction = () => {
    if (route.name !== 'Home') {
      return (
        <Icon
          onPress={() => navigation.goBack()}
          name="arrow-back"
          size={20}
          style={{ ...styles.headerIcon, ...styles.headerLeftAction }}
        />
      );
    }

    return (
      <Icon
        onPress={() => navigation.openDrawer()}
        name="menu"
        size={20}
        style={{
          ...styles.headerIcon,
          ...styles.headerLeftAction,
          color: colors.primaryText,
        }}
      />
    );
  };

  const getRightAction = () => {
    if (route.params && route.params.contact) {
      return (
        <TouchableNativeFeedback
          onPress={() => handleContactEdit(route.params.contact)}>
          <Icon
            name="mode"
            size={20}
            style={{
              ...styles.headerIcon,
              ...styles.headerEditBtn,
            }}
          />
        </TouchableNativeFeedback>
      );
    }
  };

  return {
    headerMode: 'screen',
    headerTintColor: colors.primaryText,
    headerStyle: styles.header,
    headerLeft: () => getLeftAction(),
    headerRight: () => getRightAction(),
    title: title,
  };
};

export default CustomHeader;
