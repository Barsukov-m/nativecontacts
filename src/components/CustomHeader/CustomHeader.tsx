import React from 'react';
import { View, Text } from 'react-native';
import { useNavigation, useRoute } from '@react-navigation/native';
import LeftActions from './LeftActions';
import RightActions from './RightActions';
import { NavigationParams } from './LeftActions';
import { useRemoveContactMutation } from 'store/apis/contacts';
import {
	ContactInterface,
	ContactsRouteParams,
} from '../../types/navigationTypes';
import { logWarning, wipWarning } from '../../utils/contactsUtils';
import styles from './styles';

interface CustomHeaderProps {
	title: string;
	route: {
		name: string;
		params?: {
			contact?: ContactInterface;
		};
	};
}

const CustomHeader: React.FC<CustomHeaderProps> = ({ title, route }) => {
	const { navigate } = useNavigation<NavigationParams>();

	const contact = route.params?.contact;
	const [removeContact] = useRemoveContactMutation();

	const handleEditPress = () => {
		if (!contact) {
			logWarning('Something went wrong', `Contact not found.`);
		}

		// function that could handle the contact editing
		wipWarning();
	};

	const handleDeletePress = async () => {
		if (!contact || !contact.id) {
			logWarning('Something went wrong', `Contact not found.`);
			return;
		}

		await removeContact(contact.id);
		navigate('Home');
	};

	return (
		<View style={styles.header}>
			<View style={styles.titleWrapper}>
				<LeftActions />
				<Text style={styles.title}>{title}</Text>
			</View>
			<RightActions onEdit={handleEditPress} onDelete={handleDeletePress} />
		</View>
	);
};

export default CustomHeader;
