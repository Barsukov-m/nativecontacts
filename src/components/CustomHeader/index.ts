import CustomHeader from './CustomHeader';
import LeftActions from './LeftActions';
import RightActions from './RightActions';
import styles from './styles';

export { CustomHeader, LeftActions, RightActions, styles };
