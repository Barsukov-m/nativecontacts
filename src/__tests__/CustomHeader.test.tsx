import { fireEvent, render, screen } from '@testing-library/react-native';
import { CustomHeader } from '../components/CustomHeader';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { ContactInterface } from 'types/navigationTypes';
import { wipWarning } from '../utils/contactsUtils';

jest.mock('react-native-vector-icons/MaterialIcons');

// mock contact editing function
jest.mock('../utils/contactsUtils', () => ({
	wipWarning: jest.fn(),
}));

const navigation = { goBack: jest.fn(), openDrawer: jest.fn() };

describe('<CustomHeader />', () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	it('renders correctly', () => {
		const tree = render(
			<CustomHeader
				navigation={navigation}
				route={{ name: 'Home' }}
				title="Home"
			/>
		);

		const { getByText } = tree;

		expect(getByText('Home')).toBeTruthy();

		expect(tree).toMatchSnapshot();
	});

	it('renders the left arrow icon when route name is not "Home"', () => {
		render(
			<CustomHeader
				navigation={navigation}
				route={{ name: 'ContactDetails' }}
				title="Details"
			/>
		);

		// Check for the icon name
		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'arrow-back',
			}),
			{}
		);
	});

	it('renders the menu icon when route name is "Home"', () => {
		render(
			<CustomHeader
				navigation={navigation}
				route={{ name: 'Home' }}
				title="Home"
			/>
		);

		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'menu',
			}),
			{}
		);
	});

	it('renders the edit icon when a contact parameter is provided', () => {
		const contact: ContactInterface = {
			firstName: 'Terry',
			lastName: 'Medhurst',
			phone: '+63 791 675 8914',
		};

		render(
			<CustomHeader
				navigation={navigation}
				route={{ name: 'Details', params: { contact } }}
				title="Details"
			/>
		);

		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'mode',
			}),
			{}
		);
	});

	it('triggers contact editing when the edit icon is pressed', () => {
		const contact: ContactInterface = {
			firstName: 'Terry',
			lastName: 'Medhurst',
			phone: '+63 791 675 8914',
		};

		render(
			<CustomHeader
				navigation={navigation}
				route={{ name: 'Details', params: { contact } }}
				title="Details"
			/>
		);

		const editIcon = screen.getByText(/edit-button/);

		expect(editIcon).not.toBeNull();

		if (editIcon) {
			fireEvent.press(editIcon);
		}

		expect(wipWarning).toHaveBeenCalled();
	});
});
