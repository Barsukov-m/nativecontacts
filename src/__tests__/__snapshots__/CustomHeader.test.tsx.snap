// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomHeader /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FAFAFA",
      "elevation": 5,
      "flexDirection": "row",
      "height": 56,
      "justifyContent": "space-between",
      "paddingHorizontal": 16,
      "shadowColor": "#212121",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    Icon
    <Text
      style={
        {
          "color": "#212121",
          "fontSize": 20,
          "fontWeight": "500",
        }
      }
    >
      Home
    </Text>
  </View>
</View>
`;
